packages/txservice/src/monitor/monitor.ts

```diff
@@ -0,0 +1,242 @@
+
+import { Transaction } from "./transaction";
+import { TransactionBuffer } from "./buffer";
+
+/**
+ * @classdesc Wraps and monitors transaction queue; handles transactions' initial creation and nonce assignment.
+ *
+ * All transactions created through txservice must go through here.
+ */
+export class TransactionMonitor {
+  public readonly chainId: number;
+  // This queue is used for creation of Transactions - specifically, for assigning nonce.
+  private readonly queue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+  // Buffer for monitoring transactions locally. Enables us to perform lookback and ensure all of them get through.
+  private readonly buffer: TransactionBuffer = new TransactionBuffer();
+
+  /**
+   * Centralized transaction monitoring class.
+   *
+   * @param logger The pino.BaseLogger instance we use for logging.
+   * @param provider The ChainRpcProvider instance we use for interfacing with the chain.
+   */
+  constructor(private readonly logger: BaseLogger, private readonly provider: ChainRpcProvider) {
+    this.chainId = this.provider.chainId;
+    // A separate loop will make sure they get through or get backfilled.
+    this.backfillLoop();
+  }
+
+  /**
+   * This will create a transaction with an assigned nonce as well as estimated gas / set gas price.
+   * We use this structure to essentially enforce all created transactions are saved locally in the buffer for
+   * continue monitoring, thus enabling us to further ensure they all go through.
+   * 
+   * @param minTx - Minimum transaction params needed to form a transaction for sending.
+   *
+   * @returns Transaction instance with populated params, ready for submit.
+   */
+  public async createTransaction(minTx: WriteTransaction): Promise<Transaction> {
+    // Estimate gas here will throw if the transaction is going to revert on-chain for "legit" reasons. This means
+    // that, if we get past this method, we can safely assume that the transaction will go through on submit, saving for
+    // instances where the provider malfunctions.
+    const gas = await this.getGas(minTx);
+    // Queue up the transaction with these values.
+    const result = await this.queue.add(async (): Promise<{ value: Transaction | Error; success: boolean }> => {
+      try {
+        // NOTE: This call must be here, serialized within the queue, as it is depenedent on current pending transaction count.
+        const nonce = await this.getNonce();
+        // Create a new transaction instance to track lifecycle.
+        const transaction = new Transaction(this.logger, this.provider, minTx, nonce, gas);
+        this.buffer.insert(nonce, transaction);
+        return { value: transaction, success: true };
+      } catch (e) {
+        // validate that the transaction failed on chain?
+        // didn't fail on chain? log this occurrence w/ error but ignore error ?!? -> WTF?
+        // if it did fail on chain, THROW
+        return { value: e, success: false };
+      }
```

Comment:
will this only catch the `estimateGas` errors? if so that's fine, just want to clarify

Reply:
Yes, these comments can be removed